qmail-chkpw.man
.TH qmail-chkpw 8 "" openqmail
.SH NAME
qmail-chkpw \- check a plain, login and cram-md5 SMTP authentication
.SH SYNOPSIS
.B qmail-chkpw
[
.I -e
|
.I -r
] | [
.I subprogram
] [
.I args ...
]
.SH DESCRIPTION
.B qmail-chkpw
is a 
.B checkpassword
complient interface for SMTP Authentication using
login, plain and cram-md5 authentication. It is
based on
.B cmd5checkpw-0.30
, but needs and uses encrypted passwords for AUTH PLAIN and 
AUTH LOGIN. The use of CRAM-MD5 still requires the password in clear
format. Beware that 
.B qmail-chkpw 
needs read-only access to the password file.
.SH OPTIONS
.TP
.B -e
prompt for password input, encrypt it and prints the result to stderr.
.TP 
.B -r
read a single string - usually a password - from stdin and encrypt it.

.SH FILES
.IR QMAILHOME/users/smtpauth
- this file contains pairs of logins and encrypted passwords
separated by ":". It looks like this:

login1:<encrypted password>
.br
login2:<encrypted password>
.br
login3:<plain text password>

Best way to protect it is to make it readable only for one specific user
different than your normal system users and make qmail-chkpw suid that user.
.SH RESTRICTIONS
.B qmail-chkpw
reads the file
.IR QMAILHOME/users/smtpauth 
until the first match of a user name. Without a match the authentication fails.
The user names should be unique, means that a user name can be used either with CRAM-MD5
.B or
with one of the other two authentication methods.
.SH VERSION
This documentation is about
.B qmail-chkpw
version 0.3
.SH SEE ALSO
checkpassword(8), cmd5checkpw(8)


CHANGES
20160217: moved to git, see commit messages for changes

20160130: qmail-chkpw 0.2.1
          - added '#include <crypt.h>' for linux

20160118: qmail-chkpw 0.2
          - password have to repeated with '-e' option
          - fixed compiler warnings
          - corrected config file name in man page

20151120: qmail-chkpw 0.1
          - login & plain: passwords have to be stored encrypted now
          - added option "-e" to encrypt passwords
          - fixed global.h: definitions of <inttypes.h> are used for 'UINT4'  

------------------------------------------------------------------------------
0.20
- initial public release

0.21
- accidentaly a debug version left my computer. It reads descriptor 0
instead of 3. It's fixed now.

0.22
- fix for a crash when user is not found in the passwd database. 
- Support for PLAIN/LOGIN auth (so you can use cmd5checkpw for both CRAM-MD5
and PLAIN/LOGIN). Changes contributed by Javier Kohen <jkohen@tough.com>.

0.23
- Installation customized for Qmail (Erwin Hoffmann - www.fehcom.de)

0.30
- Interface corresponds now to checkpassword:
  login & plain: userid0passwd0
  cram-md5: userid0passwd0challenge
  userid and passwd are base64 encoded; challenge is plain.


qmail-chkpw.c
#include "global.h"
#include "md5.h"
#include "hmac_md5.h"
#include <errno.h>
#include <pwd.h>
#include <stdio.h>
#include <unistd.h>

#define LINE_MAX 256
#define SASLUSERFILE "QMAILHOME/users/smtpauth"
#define READONLY "r"
char up[513];
int uplen;
static char hextab[]="0123456789abcdef";

#include "pwcrypt.c"
#include "pwmknew.c"

char *clearpw;

int doit(const char *testlogin, const char *response, unsigned char *challenge)
{
  static char line[LINE_MAX + 1];
  int found_user= 0;
  unsigned char *password = NULL;	/* 0.2: orig was 'const char */
  unsigned char digest[16];
  unsigned char digascii[33];		/* 0.2: orig was 'const char */
  unsigned char h;
  FILE *fp;
  int j;
  char *linepnt;

  int i;
  char authtype[4];

  if ((fp = fopen(SASLUSERFILE, READONLY)) == NULL) _exit(2);
  while (fgets(line, LINE_MAX, fp) != NULL) {
	if ((linepnt = strchr(line, '\n')) != NULL) {
	  *linepnt = 0;
    }
    if ((linepnt = strtok(line, ":")) == NULL) _exit(2);
    if (strcmp(linepnt, testlogin) == 0) {
	if ((linepnt = strtok(NULL, ":")) == NULL) _exit(2);
	  password = linepnt;	/* get password from file */
  	  clearpw = crypt(clearpw,linepnt);
    found_user = 1;
    break;
    }
  }
  fclose(fp);

  if (!found_user) return(1);

  /* AUTH with encrypted password */
  for (i=0;i<3;i++) { authtype[i]=password[i]; }
  authtype[4] = '\0';
  if( (authtype[0] == '$') && (authtype[0] == authtype[2]) && (authtype[1] < 1 || authtype[1] > 6) ) {
	/* password is encrypted --> no CRAM-MD5 auth */
	return(strcmp(password,clearpw)); // && strcmp(password,response));
  }
  /* do not allow plain text passwords with AUTH LOGIN/PLAIN (hard to debug!) */
  if((strcmp(clearpw,password) != 0) && ((strcmp(password,clearpw) && strcmp(password,response)) == 0)) {
	return(1);
  }

  hmac_md5( challenge, strlen(challenge), password, strlen(password), digest);

  digascii[32]=0;
  for (j=0;j<16;j++){
	h=digest[j] >> 4;
    digascii[2*j]=hextab[h];
    h=digest[j] & 0x0f;
	digascii[(2*j)+1]=hextab[h];
  }
  /* CRAM-MD5 check */
  return(strcmp(digascii,response) && strcmp(password,response));
}
/* ***** begin main ******************************************* */
int main(int argc,char **argv)
{
  char *login;
  char *response;
  char *challenge;
  int r;
  int i;
  int accepted;

  if (!argv[1]) _exit(2);

  /* process the '-e' option (interactive) */
  if(strcmp(argv[1],"-e") == 0) {
    process_e();
    _exit(0);
  }

  /* process the '-r' option (read from stdin) */
  if(strcmp(argv[1],"-r") == 0) {
	process_r();
    _exit(0);
  }

/* ********************************************************** */
  /* check for input on fd3 (the real password check)*/
  uplen = 0;
  for (;;) {
    do
      r = read(3,up + uplen,sizeof(up) - uplen);
    while ((r == -1) && (errno == EINTR));
    if (r == -1) _exit(111);
    if (r == 0) break;
    uplen += r;
    if (uplen >= sizeof(up)) _exit(1);
  }
  close(3);

  i = 0;
  login = up + i;
  while (up[i++]) if (i == uplen) _exit(2);
  response = up + i;
  if (i == uplen) _exit(2);
  while (up[i++]) if (i == uplen) _exit(2);
  challenge = up + i;

  clearpw=response; /* save original response unencrypted */

  /* some stuff for debug on command line */
/*
  printf("Input: \n");
  printf("login: %s\n",login);
  printf("resp.: %s\n",response);
  printf("chall: %s\n\n",challenge);
*/
  doencrypt(clearpw,0);		/* encrypt the response */

  accepted = doit(login,response,challenge);
  /* some stuff for debug on command line */
  printf("accepted: %i\n",accepted);	// return value of 'doit'

  for (i = 0;i < sizeof(up);++i) up[i] = 0;

  if (accepted)  _exit(1);
    execvp(argv[1],argv + 1);
    _exit(111);
}
